Example snippets for libseccomp-rs functions.

savithry-j opened the pull request:
Signed-off-by: Savithry J <Savithry.Jayaraman@sony.com>

savithry-j commented:
Dear Sugimoto san,

Could you please review the PR https://github.com/libseccomp-rs/libseccomp-rs/pull/136.

Thank you,
Savithry

rusty-snake commented on libseccomp/src/api.rs:
````diff
@@ -46,6 +54,15 @@ pub fn set_api(level: u32) -> Result<()> {
+/// ```
+/// # use libseccomp::*;
+/// set_api(1)?;
+/// assert_eq!(1, get_api()?);
````
To be consistent with the std-lib.

```suggestion
/// assert_eq!(get_api()?, 1);
```

savithry-j replied:
Added the fix in [commit](https://github.com/libseccomp-rs/libseccomp-rs/commit/2e07e17f07f92e2b6f7766c3c2d67495dc762efe)

rusty-snake commented on libseccomp/src/filter_context.rs:
````diff
@@ -432,6 +493,16 @@ impl ScmpFilterContext {
+    /// ```
+    /// # use libseccomp::*;
+    /// let mut ctx = ScmpFilterContext::new_filter(ScmpAction::Allow)?;
+    /// ctx.set_ctl_nnp(true)?;
+    /// assert!(ctx.get_ctl_nnp()?);
````
Since `true` is the default I would flip the logic:
```suggestion
    /// ctx.set_ctl_nnp(false)?;
    /// assert!(!ctx.get_ctl_nnp()?);
```

savithry-j replied:
Added the fix in [commit](https://github.com/libseccomp-rs/libseccomp-rs/commit/2e07e17f07f92e2b6f7766c3c2d67495dc762efe)

rusty-snake commented on libseccomp/src/filter_context.rs:
````diff
@@ -594,6 +730,15 @@ impl ScmpFilterContext {
+    /// ```
+    /// # use libseccomp::*;
+    /// let mut ctx = ScmpFilterContext::new_filter(ScmpAction::Allow)?;
+    /// ctx.set_ctl_nnp(true)?;
````
```suggestion
    /// ctx.set_ctl_nnp(false)?;
```

savithry-j replied:
Added the fix in [commit](https://github.com/libseccomp-rs/libseccomp-rs/commit/2e07e17f07f92e2b6f7766c3c2d67495dc762efe)

rusty-snake commented on libseccomp/src/filter_context.rs:
```diff
@@ -487,6 +578,21 @@ impl ScmpFilterContext {
+    /// # use libseccomp::*;
+    /// let mut ctx = ScmpFilterContext::new_filter(ScmpAction::Allow)?;
+    /// if check_api(4, ScmpVersion::from((2, 5, 0))).unwrap() {
+    ///     ctx.set_ctl_ssb(true)?;
+    ///     assert!(ctx.get_ctl_ssb()?);
+    /// } else {
+    ///     assert!(ctx.set_ctl_ssb(true).is_err());
+    ///     assert!(ctx.get_ctl_ssb().is_err());
```
`true` is the default IIRC

```suggestion
    ///     ctx.set_ctl_ssb(false)?;
    ///     assert!(!ctx.get_ctl_ssb()?);
    /// } else {
    ///     assert!(ctx.set_ctl_ssb(false).is_err());
    ///     assert!(ctx.get_ctl_ssb().is_err());
```

savithry-j replied:
Added the fix in [commit](https://github.com/libseccomp-rs/libseccomp-rs/commit/2e07e17f07f92e2b6f7766c3c2d67495dc762efe)

rusty-snake commented on libseccomp/src/filter_context.rs:
````diff
@@ -670,6 +833,19 @@ impl ScmpFilterContext {
+    /// ```
+    /// # use libseccomp::*;
+    /// let mut ctx = ScmpFilterContext::new_filter(ScmpAction::Allow)?;
+    /// if check_api(4, ScmpVersion::from((2, 5, 0))).unwrap() {
+    ///     ctx.set_ctl_ssb(true)?;
+    /// } else {
+    ///     assert!(ctx.set_ctl_ssb(true).is_err());
+    /// }
````
```suggestion
    /// if check_api(4, ScmpVersion::from((2, 5, 0))).unwrap() {
    ///     ctx.set_ctl_ssb(false)?;
    /// } else {
    ///     assert!(ctx.set_ctl_ssb(false).is_err());
    /// }
```

savithry-j replied:
Added the fix in [commit](https://github.com/libseccomp-rs/libseccomp-rs/commit/2e07e17f07f92e2b6f7766c3c2d67495dc762efe)

rusty-snake reviewed (commented):
Looks good.

savithry-j commented:
Dear @rusty-snake san and @ManaSugi san,

Thank you for the review.
Added the suggested changes.

Thank you
Savithry

ManaSugi commented on libseccomp/src/filter_context.rs:
````diff
@@ -541,6 +632,21 @@ impl ScmpFilterContext {
+    /// ```
+    /// # use libseccomp::*;
+    /// let mut ctx = ScmpFilterContext::new_filter(ScmpAction::Allow)?;
+    /// if check_api(4, ScmpVersion::from((2, 5, 0))).unwrap() {
````
Could you hide these branches using `#` ?
I think it will be good example code if you can hide the branch.
The same is true of other example code.

ManaSugi commented:
@savithry-j LGTM, Thank you for great contribution!
@rusty-snake, If you are Ok, could you approve and merge this?

savithry-j commented:
> @savithry-j Please squash the commits?

Yes, I have squashed them now. Thank you

rusty-snake commented on libseccomp/src/filter_context.rs:
````diff
@@ -842,6 +1054,16 @@ impl ScmpFilterContext {
+    ///
+    /// ```
+    /// # use libseccomp::*;
+    /// # use std::io::{stdout, Error};
````
Do we need `Error`? I don't see it used.

savithry-j replied:
@rusty-snake, I have removed it, Thank you for reviewing.

rusty-snake commented on libseccomp/src/filter_context.rs:
````diff
@@ -824,6 +1026,16 @@ impl ScmpFilterContext {
+    ///
+    /// ```
+    /// # use libseccomp::*;
+    /// # use std::io::{stdout, Error};
````
s.a.

savithry-j replied:
@rusty-snake, I have removed it, Thank you for reviewing.

rusty-snake reviewed (commented):
almost approved

savithry-j commented:
@rusty-snake could you please review for further changes?

ManaSugi commented:
@savithry-j Please squash the commits.

rusty-snake commented on CHANGELOG.md:
```diff
@@ -31,6 +31,9 @@ added/removed (**Incompatible change**).
 ### Fixed
 - `scmp_cmp!`: `allow(unused_parens)` in `$mask`
 
+### Fixed
+- `scmp_cmp!`: `allow(unused_parens)` in `$mask`
+
```
;)

ManaSugi commented on libseccomp/src/filter_context.rs:
```diff
@@ -538,6 +629,21 @@ impl ScmpFilterContext {
+    /// if check_api(4, ScmpVersion::from((2, 5, 0))).unwrap() {
+    ///     ctx.set_ctl_ssb(false)?;
+    ///     assert!(!ctx.get_ctl_ssb()?);
+    /// } else {
+    ///     assert!(ctx.set_ctl_ssb(false).is_err());
+    ///     assert!(ctx.get_ctl_ssb().is_err());
+    /// }
```
This else is not necessary in example code.

rusty-snake replied:
My idea (about negative out come because of version/api-level) for this case as an exmaple:
```rust
    /// # use libseccomp::*;
    /// let mut ctx = ScmpFilterContext::new_filter(ScmpAction::Allow)?;
    /// # if check_api(4, ScmpVersion::from((2, 5, 0))).unwrap() {
    /// ctx.set_ctl_ssb(false)?;
    /// assert!(!ctx.get_ctl_ssb()?);
    /// # }
    /// # Ok::<(), Box<dyn std::error::Error>>(())
```

ManaSugi commented on libseccomp/src/filter_context.rs:
```diff
@@ -559,6 +665,21 @@ impl ScmpFilterContext {
+    /// if check_api(4, ScmpVersion::from((2, 5, 0)))? {
+    ///     ctx.set_ctl_optimize(2)?;
+    ///     assert_eq!(ctx.get_ctl_optimize()?, 2);
+    /// } else {
+    ///     assert!(ctx.set_ctl_optimize(2).is_err());
+    ///     assert!(ctx.get_ctl_optimize().is_err());
+    /// }
```
This else is not necessary in example code.

ManaSugi commented on libseccomp/src/filter_context.rs:
```diff
@@ -580,6 +701,21 @@ impl ScmpFilterContext {
+    /// if check_api(4, ScmpVersion::from((2, 5, 0)))? {
+    ///     ctx.set_api_sysrawrc(true)?;
+    ///     assert!(ctx.get_api_sysrawrc()?);
+    /// } else {
+    ///     assert!(ctx.set_api_sysrawrc(true).is_err());
+    ///     assert!(ctx.get_api_sysrawrc().is_err());
+    /// }
```
This else is not necessary in example code.

ManaSugi commented on libseccomp/src/filter_context.rs:
```diff
@@ -745,6 +908,19 @@ impl ScmpFilterContext {
+    /// let mut ctx = ScmpFilterContext::new_filter(ScmpAction::Allow)?;
+    /// if check_api(4, ScmpVersion::from((2, 5, 0))).unwrap() {
+    ///     ctx.set_ctl_ssb(false)?;
+    /// } else {
+    ///     assert!(ctx.set_ctl_ssb(false).is_err());
+    /// }
```
This else is not necessary in example code.

ManaSugi commented on libseccomp/src/filter_context.rs:
```diff
@@ -777,6 +953,19 @@ impl ScmpFilterContext {
+    /// let mut ctx = ScmpFilterContext::new_filter(ScmpAction::Allow)?;
+    /// if check_api(4, ScmpVersion::from((2, 5, 0)))? {
+    ///     ctx.set_ctl_optimize(2)?;
+    /// } else {
+    ///     assert!(ctx.set_ctl_optimize(2).is_err());
+    /// }
```
This else is not necessary in example code.

ManaSugi commented on libseccomp/src/filter_context.rs:
```diff
@@ -803,6 +992,19 @@ impl ScmpFilterContext {
+    /// let mut ctx = ScmpFilterContext::new_filter(ScmpAction::Allow)?;
+    /// if check_api(4, ScmpVersion::from((2, 5, 0)))? {
+    ///     ctx.set_api_sysrawrc(true)?;
+    /// } else {
+    ///     assert!(ctx.set_api_sysrawrc(true).is_err());
+    /// }
```
This else is not necessary in example code.

ManaSugi commented on libseccomp/src/functions.rs:
```diff
@@ -24,6 +24,18 @@ use libseccomp_sys::*;
+/// # use libseccomp::*;
+/// if check_version(ScmpVersion::from((2, 5, 1)))? {
+///     assert!(reset_global_state().is_ok());
+/// } else {
+///     assert!(reset_global_state().is_err());
+/// }
```
This else is not necessary in example code.

ManaSugi commented on libseccomp/src/functions.rs:
````diff
@@ -24,6 +24,18 @@ use libseccomp_sys::*;
+/// ```
+/// # use libseccomp::*;
+/// if check_version(ScmpVersion::from((2, 5, 1)))? {
+///     assert!(reset_global_state().is_ok());
````
You can remove assert macro.

rusty-snake replied:
```suggestion
///     reset_global_state()?;
```

ManaSugi reviewed (commented):
Could you unnecessary assert macro?
Example code is not test code, so I'd like to remove assert macro as much as possible for making the example clearer.

rusty-snake commented:
> Example code is not test code

Example code is tested by `cargo test` for library targets.

rusty-snake commented:
> so I'd like to remove assert macro as much as possible for making the example clearer.

How should we then show the expected function result? With an comment? While it works with comments, the can become outdated (see above).

ManaSugi commented:
> Example code is tested by `cargo test` for library targets. This is done to make sure that the example is updated too if you update the function (i.e. you don't get outdated examples).

@rusty-snake Thank you for the comment.
Yes, but the `assert` I requested to remove are not always necessary like the existing examples of  `add_arch`, `add_rule`, and etc. If we want to test all example codes not to get them outdated, we will have to add `assert` to the existing examples too strictly. But actually I wonder we need to add `assert` to all examples.  Does it make sense to hide `assert` from example codes by comment out? In that case, we can check the examples with keeping them clean, If you know the best way of writing good examples, please let me know.

rusty-snake commented on libseccomp/src/filter_context.rs:
```diff
@@ -839,6 +1051,16 @@ impl ScmpFilterContext {
+    /// # use libseccomp::*;
+    /// # use std::io::{stdout};
+    /// let ctx = ScmpFilterContext::new_filter(ScmpAction::Allow)?;
+    /// assert!(ctx.export_bpf(&mut stdout()).is_ok());
```
```suggestion
    /// ctx.export_bpf(&mut stdout())?;
```

rusty-snake commented on libseccomp/src/filter_context.rs:
```diff
@@ -821,6 +1023,16 @@ impl ScmpFilterContext {
+    /// # use libseccomp::*;
+    /// # use std::io::{stdout};
+    /// let ctx = ScmpFilterContext::new_filter(ScmpAction::Allow)?;
+    /// assert!(ctx.export_pfc(&mut stdout()).is_ok());
```
```suggestion
    /// ctx.export_pfc(&mut stdout())?;
```

rusty-snake commented:
> If we want to test all example codes not to get them outdated, we will have to add assert to the existing examples too strictly.

Yes. But this can be done over time.

> Does it make sense to hide assert from example codes by comment out?

Yes, it think this is a good solution to make sure examples get updated while hiding unnecessary details.

> If you know the best way of writing good examples, please let me know.

For "unit-examples" on functions w/o side-effects this is easy (just use `assert_eq!`). However the most functions in libseccomp are called because of thier side-effects. For such functions it's difficult to make an example which is clear/simple and shows the usage of the function.

For the asserts you marked I think we can come to
- positive out come should use the question mark operator anyway.
- negative out come because of libseccomp version/api-level can be ignored. Maybe we can even hide the version/api-level check.

ManaSugi commented:
> For "unit-examples" on functions w/o side-effects this is easy (just use assert_eq!). However the most functions in libseccomp are called because of thier side-effects. For such functions it's difficult to make an example which is clear/simple and shows the usage of the function.

I completely agree. 

> For the asserts you marked I think we can come to
> positive out come should use the question mark operator anyway.
> negative out come because of libseccomp version/api-level can be ignored. Maybe we can even hide the version/api-level > check.

Yes, if something you have @savithry-j to fix, please suggests them. Thank you
@savithry-j Could you take a look the comments @rusty-snake mentioned?

savithry-j commented:
Dear @rusty-snake and @ManaSugi san,

Thank you for the suggestions. I checked few known crates for reference:
- https://docs.rs/nom/latest/nom/
- https://api.rocket.rs/v0.5-rc/rocket/catcher/struct.Catcher.html
- https://docs.rs/textwrap/0.15.0/textwrap/
- https://docs.rs/reqwest/0.11.10/reqwest/struct.RequestBuilder.html
- https://thume.ca/crates/

They use assert macro in the example codes of functions.

For reference, [try_clone](https://docs.rs/reqwest/0.11.10/reqwest/struct.RequestBuilder.html#method.try_clone).

Hence shall we keep the assert macro for clear understanding of the usage of function for now? 
Later, when we find a good documentation/reference on the standard of writing examples,  shall we revisit these changes?

Also, addressed the review comments. Could you please review these latest modifications in the examples.

Thank you,
Savithry

rusty-snake commented:
> Hence shall we keep the assert macro for clear understanding of the usage of function for now?

I say yes.

Conditions would be
- there is anything useful to assert.
- it does not assert `Ok(_)` (this should be done with `?`).
- it does not assert `Err(_)` because of version/api-level.

ManaSugi reviewed (approved):
LGTM, these examples are good.
@savithry-j Thank you for investigating how other crates write examples.
@rusty-snake Thank you for the great review and comments!
